File: 1/2020.09.28/46.Permutations/python/solution.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def backtracking(res):
            # terminator
            if len(res) == len(nums):
                self.res.append(res[:])
                return
            # process current level
            for num in nums:
                if num not in res:
                    res.append(num)
                    # drill down
                    backtracking(res)
                    # reset states
                    res.pop()
        self.res = []
        backtracking([])
        return self.res

File: 1/2020.09.28/46.Permutations/java/solution.java
class Solution {

    private List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        backtracking(new ArrayList<>(), nums);
        return res;
    }

    private void backtracking(List<Integer> tmpList, int[] nums) {
        // terminator
        if (tmpList.size() == nums.length) {
            res.add(new ArrayList<>(tmpList));
            return;
        }
        // process current level
        for (int num : nums) {
            if (!tmpList.contains(num)) {
                tmpList.add(num);
                // drill down
                backtracking(tmpList, nums);
                // reset states
                tmpList.remove(tmpList.size() - 1);
            }
        }
    }
}
